### PR TITLE
Force the residual vector to be in the range of OP. (#79)

### PR DESCRIPTION
**Pull request by caliarim**
Pull request for issue (#79) (force the initial residual to be in the range of the operator).

### CODE REVIEW

**Comment by sylvestre** [comment]
Many thanks for your PR!

Seems that this is failing:
https://travis-ci.org/opencollab/arpack-ng/jobs/333185780
The following tests FAILED:
	 77 - bug_79_double_complex (Not Run)

**Comment by caliarim** [comment]
I forgot to add the test to add_dependencies in CMakeLists.txt?

**Comment by sylvestre** [comment]
@turboencabulator  @thrasibule
any option on this?

**Comment by turboencabulator** [comment]
Build script changes look good to me.

File: SRC/cgetv0.f [hunk 1]
```diff
@@ -271,9 +269,9 @@ subroutine cgetv0
 c
       call arscnd (t2)
       first = .TRUE.
+      call ccopy (n, workd(n+1), 1, resid, 1)
```
Comment (thrasibule) [inline SRC/cgetv0.f]: Why is this needed? We only end up here if ido != 0.

**Comment by thrasibule** [comment]
This seems pretty low risk, and it fixes a test case, so looks good to me.

File: SRC/cgetv0.f (hunk 1 above)
Comment (caliarim) [reply SRC/cgetv0.f]: I'm away from my laptop, but aftet the very first OP application ido is -1, isn't it? So we need to copy the result of OP application back to resid.

**Comment by caliarim** [comment]
Shall we change the documentation somewhere about this change? For instance, in X{s,n}aupd it is written

> c  RESID   Complex*16  array of length N.  (INPUT/OUTPUT)
> c          On INPUT:
> c          If INFO .EQ. 0, a random initial residual vector is used.
> c          If INFO .NE. 0, RESID contains the initial residual vector,
> c                          possibly from a previous run.

No mention to the range of the operator. Basically the same in X{s,n}aup2. In Xgetv0 it is written

> c  Force the residual vector to be in the range of the operator OP.

In the book, page 72, it is written

> In either case, it the (shift-invert) computational mode calls for it, the vector is also forced to be in the range of OP.

But with this change the vector is forced to be in the range for the regular mode, too. And it has always been forced to be in the range for the generalized problem.
And what happens if somebody really would like to test her/his own inital vector?

**Comment by sylvestre** [comment]
@caliarim is that ready to be merged? Thanks

**Comment by caliarim** [comment]
I think so. With respect to my previous comment, in Xgetv0 it is written that the initial vector is in the range, and now this is (always) true. And of course we cannot change the book...

**Comment by sylvestre** [comment]
thanks!

**Comment by ViralBShah** [comment]
Apologies for chiming in on such an old issue. We believe this PR is causing the Arpack tests in Julia to fail. The Julia test case is described in https://github.com/JuliaLinearAlgebra/Arpack.jl/issues/118, and fails once in a while when run repeatedly.

It may very well be an issue in our interface, but this worked reliably in Arpack 3.5, and not in any of the other later releases. In case someone here notices something we are doing wrong, or has some insight - it would be great.

**Comment by ViralBShah** [comment]
Apologies for chiming in on a closed PR. For Julia's Arpack.jl, we believe that this has introduced a bug (it is our best guess):

The test case in https://github.com/JuliaLinearAlgebra/Arpack.jl/issues/118 demonstrates the following behaviour:
1. Fails about 2% of the time with arpack-ng 3.5. If initialized with a random `v0`, it passes reliably.
2. Fails always on later versions of arpack-ng, including 3.8, even when initialized with a random `v0` (demonstrated in https://github.com/JuliaLinearAlgebra/Arpack.jl/pull/145)

**Comment by sylvestre** [comment]
@ViralBShah nice to see you there and no worries for this.

@caliarim do you want to propose a PR ? :)
thanks

**Comment by fghoussen** [comment]
@ViralBShah: could revert both d04bdf19b9189a971f8c3bf2561b8058994e9cb4 and 31854cadaff067e30b8715f10533c459a9b7d447 related commits and check things are getting better at your side? Making sure the reg you see comes from that

AFAIU, this would imply reopening #79 and #142!...
